rfc/20241206-oci-registries/9-auth-implementation-details.md

```diff
@@ -0,0 +1,64 @@
+
+| [« Previous](8-open-questions.md) | [Up](../20241206-oci-registries.md) | [Next »](10-provider-implementation-details.md) |
+
+---
+
+This appendix discusses implementation details related to [OCI Registry Authentication](7-authentication.md).
+
+> [!WARNING]
+> This appendix is still under construction, subject to change based on feedback on the earlier chapters, and may not yet be up-to-date with the latest changes in the earlier chapters.
+
+## Registry authentication is a cross-cutting concern
+
+While most of what we've discussed in this RFC is defined separately for provider and module package installation, the _authentication_ implementation should be shared between both and designed so that it could potentially be used for any other OCI registry interactions we might implement in future, such as hypothetical support for storing OpenTofu state snapshots as OCI artifacts.
+
+Therefore our primary concern for implementation is in centrally-managing the credentials settings and then making them available to the relevant components of both the provider and module package installers.
+
+## OpenTofu CLI Configuration and `package main`
+
+Although there is considerable existing legacy code not following this pattern, the current intended design for OpenTofu is to follow the [dependency inversion principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) with `package main` acting as the ultimate arbiter of how different subsystems are configured to work together. The main package in turn uses `package cliconfig` (`internal/command/cliconfig`) to decide most of the locally-user-configurable settings that can affect those dependency resolution decisions.
+
+We will continue that design approach by teaching `package cliconfig` to decode and validate the `oci` block in the CLI configuration, returning the discovered information as part of the overall `cliconfig.Config` object generated by that package.
+
+The implicit configuration mode acts as an alternative way to populate the same settings from Docker CLI or other container system configuration files, and so would also be implemented in `package cliconfig` by mapping the concepts from the Docker CLI configuration language to the same internal data types that we would decode our explicit configuration into, so that the rest of the system does not need to be concerned about how that information was discovered.
+
+`package main` is responsible for using the information returned from the CLI configuration to configure and instantiate the [`ociclient.OCIClient`](https://pkg.go.dev/github.com/opentofu/libregistry@v0.0.0-20241121135917-6f06a9a60bb5/registryprotocols/ociclient#OCIClient) that will then be passed as a dependency into both the provider installer and the module installer, which will then encapsulate all of the OCI registry interactions including the collection and inclusion of credentials when making requests.
```

Comment:
I'd rather keep the initial implementation of this within the opentofu core.  Deciding how to break out "getprovider/module" functionality into a library like libregistry is a much larger conversation that I believe should happen in a separate location.

I see the utility of moving in that direction, but want to make sure that we consider API design more holistically and not just carve out "This is how we do OCI providers".

Reply:
FWIW it seems like I just yoinked this out of the earlier RFC that was based on the prototype we built:

https://github.com/opentofu/opentofu/blob/b6f6349945d1a73a95efd46fa0b401c70322511d/rfc/20241206-oci-registry-auth-config.md?plain=1#L93

We'd copied this text over here with the intention of revising it once the end-user-facing portion of the RFC text had settled, but I didn't actually return and do that and so this section in particular is still a little "off" in various ways compared to the user-facing portion. We'll presumably want to revise this more heavily once we've finalized the user-facing design and made some specific decisions about how to approach the implementation.